File: week1-cpp/Application_Sunglasses_Filter/Using_Arithmetic_Operations.cpp
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/opencv.hpp>
#include "dataPath.hpp"

using namespace std;
using namespace cv;

int main(void){
	// Make the dimensions of the mask same as the input image.
	// Since Face Image is a 3-channel image, we create a 3 channel image for the mask
	Mat glassMask1 = imread("../results/sunglassAlpha.png",0);
	Mat glassBGR = imread("../results/sunglassRGB.png");
	Mat glassMask;
	Mat glassMaskChannels[] = {glassMask1,glassMask1,glassMask1};
	merge(glassMaskChannels,3,glassMask);
	// Make the values [0,1] since we are using arithmetic operations
	glassMask = glassMask/255;

    // Load the Face Image
    string faceImagePath = DATA_PATH + "/images/musk.jpg";
    Mat faceImage = imread(faceImagePath);
	
	// Make a copy
	Mat faceWithGlassesArithmetic = faceImage.clone();

	// Get the eye region from the face image
	Mat eyeROI = faceWithGlassesArithmetic(Range(150,250),Range(140,440));

	Mat eyeROIChannels[3];
	split(eyeROI,eyeROIChannels);
	Mat maskedEyeChannels[3];
	Mat maskedEye;

	for (int i = 0; i < 3; i++)
	{
	    // Use the mask to create the masked eye region
	    multiply(eyeROIChannels[i], (1-glassMaskChannels[i]), maskedEyeChannels[i]);
	}

	merge(maskedEyeChannels,3,maskedEye);

	Mat maskedGlass;
	// Use the mask to create the masked sunglass region
	multiply(glassBGR, glassMask, maskedGlass);
	
	Mat eyeRoiFinal;
	// Combine the Sunglass in the Eye Region to get the augmented image
	add(maskedEye, maskedGlass, eyeRoiFinal);

	imwrite("../results/maskedEyeRegion.png",maskedEye);
	imwrite("../results/maskedSunglassRegion.png",maskedGlass);
	imwrite("../results/augmentedEyeAndSunglass.png",eyeRoiFinal);

	//imshow("Masked eye region", maskedEye);
	//imshow("Masked Sunglass Region", maskedGlass);
	//imshow("Augmented Eye and Sunglass", eyeRoiFinal);
	//waitKey(0);
	
	// Replace the eye ROI with the output from the previous section
	eyeRoiFinal.copyTo(eyeROI);

	imwrite("../results/withSunglasses.png",faceWithGlassesArithmetic);

	//imshow("With Sunglasses", faceWithGlassesArithmetic);
	//waitKey(0);
	
	return 0;
}

File: week1-python/Application_Sunglasses_Filter/Using_Arithmetic_Operations.py
# Import libraries
import cv2
import numpy as np
from dataPath import DATA_PATH

# Load the Face Image
faceImagePath = DATA_PATH + "/images/musk.jpg"
faceImage = cv2.imread(faceImagePath)

#cv2.imshow("Face",faceImage)

# Load the Sunglass image with Alpha channel
# (http://pluspng.com/sunglass-png-1104.html)
glassimagePath = DATA_PATH + "/images/sunglass.png"
glassPNG = cv2.imread(glassimagePath,-1)

# Resize the image to fit over the eye region
glassPNG = cv2.resize(glassPNG,(300,100))
print("image Dimension ={}".format(glassPNG.shape))

# Separate the Color and alpha channels
glassBGR = glassPNG[:,:,0:3]
glassMask1 = glassPNG[:,:,3]

# Make the dimensions of the mask same as the input image.
# Since Face Image is a 3-channel image, we create a 3 channel image for the mask
glassMask = cv2.merge((glassMask1,glassMask1,glassMask1))

# Make the values [0,1] since we are using arithmetic operations
glassMask = np.uint8(glassMask/255)

# Make a copy
faceWithGlassesArithmetic = faceImage.copy()

# Get the eye region from the face image
eyeROI= faceWithGlassesArithmetic[150:250,140:440]

# Use the mask to create the masked eye region
maskedEye = cv2.multiply(eyeROI,(1-  glassMask ))

# Use the mask to create the masked sunglass region
maskedGlass = cv2.multiply(glassBGR,glassMask)

# Combine the Sunglass in the Eye Region to get the augmented image
eyeRoiFinal = cv2.add(maskedEye, maskedGlass)

#cv2.imshow("Masked Eye Region",maskedEye)
#cv2.imshow("Masked Sunglass Region",maskedGlass)
#cv2.imshow("Augmented Eye and Sunglass",eyeRoiFinal)
cv2.imwrite("../results/maskedEyeRegion.png",maskedEye)
cv2.imwrite("../results/maskedSunglassRegion.png",maskedGlass)
cv2.imwrite("../results/augmentedEyeAndSunglass.png",eyeRoiFinal)

# Replace the eye ROI with the output from the previous section
faceWithGlassesArithmetic[150:250,140:440]=eyeRoiFinal

#cv2.imshow("With Sunglasses",faceWithGlassesArithmetic)
#cv2.waitKey(0)
#cv2.destroyAllWindows()
cv2.imwrite("../results/withSunglasses.png",faceWithGlassesArithmetic)
